File: apps/mobile/android/app/src/main/java/com/houselevi/plus/viewmodel/ShopViewModel.kt
package com.houselevi.plus.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.houselevi.plus.data.api.HLApiClient
import com.houselevi.plus.data.models.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

//  Shop Home 
data class ShopHomeState(
    val isLoading: Boolean        = true,
    val products: List<Product>   = emptyList(),
    val error: String?            = null,
)

class ShopHomeViewModel : ViewModel() {
    private val _state = MutableStateFlow(ShopHomeState())
    val state: StateFlow<ShopHomeState> = _state
    init { load() }

    fun load() {
        viewModelScope.launch {
            _state.value = ShopHomeState(isLoading = true)
            try {
                val res = HLApiClient.shopApi.getProducts()
                _state.value = ShopHomeState(isLoading = false, products = res.body()?.data ?: emptyList())
            } catch (e: Exception) {
                _state.value = ShopHomeState(isLoading = false, error = "Failed to load products.")
            }
        }
    }
}

//  Product Detail 
data class ProductDetailState(
    val isLoading: Boolean          = true,
    val product: Product?           = null,
    val selectedVariant: ProductVariant? = null,
    val quantity: Int               = 1,
    val addedToCart: Boolean        = false,
    val error: String?              = null,
)

class ProductDetailViewModel : ViewModel() {
    private val _state = MutableStateFlow(ProductDetailState())
    val state: StateFlow<ProductDetailState> = _state

    fun load(productId: String) {
        viewModelScope.launch {
            _state.value = ProductDetailState(isLoading = true)
            try {
                val res = HLApiClient.shopApi.getProductById(productId)
                val product = res.body()?.data
                _state.value = ProductDetailState(
                    isLoading = false,
                    product = product,
                    selectedVariant = product?.variants?.firstOrNull { it.inStock }
                        ?: product?.variants?.firstOrNull(),
                )
            } catch (e: Exception) {
                _state.value = ProductDetailState(isLoading = false, error = "Failed to load product.")
            }
        }
    }

    fun selectVariant(v: ProductVariant) {
        _state.value = _state.value.copy(selectedVariant = v, addedToCart = false)
    }

    fun setQuantity(q: Int) {
        _state.value = _state.value.copy(quantity = q.coerceIn(1, 10))
    }

    fun addToCart(auth: String) {
        val state = _state.value
        val product = state.product ?: return
        val variant = state.selectedVariant ?: return
        viewModelScope.launch {
            try {
                HLApiClient.shopApi.addToCart(
                    auth = "Bearer $auth",
                    body = AddToCartBody(
                        productId    = product.id,
                        variantId    = variant.id,
                        variantTitle = variant.title,
                        price        = variant.price.takeIf { it > 0 } ?: product.basePrice,
                        quantity     = state.quantity,
                    )
                )
                _state.value = _state.value.copy(addedToCart = true)
            } catch (_: Exception) {}
        }
    }
}

//  Cart 
data class CartState(
    val isLoading: Boolean  = true,
    val cart: Cart          = Cart(),
    val checkoutUrl: String? = null,
    val error: String?      = null,
)

class CartViewModel : ViewModel() {
    private val _state = MutableStateFlow(CartState())
    val state: StateFlow<CartState> = _state

    fun load(auth: String) {
        viewModelScope.launch {
            _state.value = CartState(isLoading = true)
            try {
                val res = HLApiClient.shopApi.getCart("Bearer $auth")
                _state.value = CartState(isLoading = false, cart = res.body()?.cart ?: Cart())
            } catch (e: Exception) {
                _state.value = CartState(isLoading = false, error = "Failed to load cart.")
            }
        }
    }

    fun updateQuantity(auth: String, itemId: String, qty: Int) {
        viewModelScope.launch {
            try {
                val res = HLApiClient.shopApi.updateQuantity("Bearer $auth", itemId, UpdateQuantityBody(qty))
                res.body()?.cart?.let { _state.value = _state.value.copy(cart = it) }
            } catch (_: Exception) {}
        }
    }

    fun removeItem(auth: String, itemId: String) {
        viewModelScope.launch {
            try {
                val res = HLApiClient.shopApi.removeItem("Bearer $auth", itemId)
                res.body()?.cart?.let { _state.value = _state.value.copy(cart = it) }
            } catch (_: Exception) {}
        }
    }

    fun checkout(auth: String, phone: String, address: ShippingAddress, location: String) {
        viewModelScope.launch {
            try {
                val res = HLApiClient.shopApi.createOrder(
                    "Bearer $auth",
                    CreateOrderBody(phone = phone, shippingAddress = address, shippingLocation = location)
                )
                val url = res.body()?.pesapalUrl
                if (!url.isNullOrBlank()) _state.value = _state.value.copy(checkoutUrl = url)
                else _state.value = _state.value.copy(error = "Checkout failed. Please try again.")
            } catch (e: Exception) {
                _state.value = _state.value.copy(error = "Network error. Please try again.")
            }
        }
    }

    fun clearCheckoutUrl() { _state.value = _state.value.copy(checkoutUrl = null) }
}

File: apps/mobile/ios/HouseLevi+/ViewModels/ShopViewModel.swift
import Foundation

@MainActor
final class ShopHomeViewModel: ObservableObject {
    @Published var products: [Product] = []
    @Published var isLoading = true
    @Published var error: String? = nil
    init() { Task { await load() } }
    func load() async {
        isLoading = true; error = nil
        do { products = try await HLAPIClient.shared.getProducts().data }
        catch { self.error = "Failed to load products." }
        isLoading = false
    }
}

@MainActor
final class ProductDetailViewModel: ObservableObject {
    @Published var product: Product? = nil
    @Published var selectedVariant: ProductVariant? = nil
    @Published var quantity = 1
    @Published var isLoading = true
    @Published var addedToCart = false
    @Published var error: String? = nil

    func load(productId: String) async {
        isLoading = true
        do {
            product = try await HLAPIClient.shared.getProductById(productId).data
            selectedVariant = product?.variants.first(where: { $0.inStock }) ?? product?.variants.first
        } catch { self.error = "Failed to load product." }
        isLoading = false
    }
    func selectVariant(_ v: ProductVariant) { selectedVariant = v; addedToCart = false }
    func setQuantity(_ q: Int) { quantity = max(1, min(q, 10)) }

    func addToCart(auth: String) async {
        guard let p = product, let v = selectedVariant else { return }
        do {
            _ = try await HLAPIClient.shared.addToCart(auth: auth, body: AddToCartBody(
                productId: p.id, variantId: v.id, variantTitle: v.title,
                price: v.price > 0 ? v.price : p.basePrice, quantity: quantity))
            addedToCart = true
        } catch {}
    }
}

@MainActor
final class CartViewModel: ObservableObject {
    @Published var cart = Cart()
    @Published var isLoading = true
    @Published var checkoutUrl: String? = nil
    @Published var error: String? = nil

    func load(auth: String) async {
        isLoading = true
        do { cart = try await HLAPIClient.shared.getCart(auth: auth).cart }
        catch { self.error = "Failed to load cart." }
        isLoading = false
    }
    func updateQuantity(auth: String, itemId: String, qty: Int) async {
        do { cart = try await HLAPIClient.shared.updateCartQuantity(auth: auth, itemId: itemId, body: UpdateQuantityBody(quantity: qty)).cart }
        catch {}
    }
    func removeItem(auth: String, itemId: String) async {
        do { cart = try await HLAPIClient.shared.removeCartItem(auth: auth, itemId: itemId).cart }
        catch {}
    }
    func checkout(auth: String, phone: String, address: ShippingAddress, location: String) async {
        do {
            let res = try await HLAPIClient.shared.createOrder(auth: auth, body: CreateOrderBody(
                phone: phone, shippingAddress: address, shippingLocation: location))
            if !res.pesapalUrl.isEmpty { checkoutUrl = res.pesapalUrl }
            else { error = "Checkout failed. Please try again." }
        } catch { self.error = "Network error. Please try again." }
    }
}
